File: src/main/java/com/mikea/gae/rx/RxPushStream.java
package com.mikea.gae.rx;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class RxPushStream<T> extends RxStream<T> {
    private List<IObserver<T>> observers = new ArrayList<>();

    protected RxPushStream(Rx rx) {
        super(rx);
    }

    public synchronized void onNext(T t) throws IOException {
        for (IObserver<T> observer : observers) {
            observer.onNext(t);
        }
    }

    @Override
    public synchronized IDisposable subscribe(IObserver<T> observer) {
        observers.add(observer);
        return new IDisposable() {
            @Override
            public void dispose() {
                throw new UnsupportedOperationException();
            }
        };
    }
}

File: src/main/scala/com/mikea/gae/rx/RxPushStream.scala
package com.mikea.gae.rx

import com.mikea.gae.rx.base.{IObservable, IDisposable, IObserver}
import com.google.inject.Injector

class RxPushStream[T](_injector: Injector) extends IObservable[T] {
  def onNext(t: T): Unit = {
    for (observer <- observers) {
      observer.onNext(t)
    }
  }

  def subscribe(observer: IObserver[T]): IDisposable = {
    observers = observers :+ observer

    new IDisposable {
      def dispose(): Unit = {
        throw new UnsupportedOperationException
      }
    }
  }

  private var observers: Vector[IObserver[T]] = Vector()

  def instantiate[C](aClass: Class[C]) = _injector.getInstance(aClass)
}